File: lecture/inflearn-spring-boot/autoconfig/src/main/java/hello/config/DbConfig.java
package hello.config;

import com.zaxxer.hikari.HikariDataSource;
import javax.sql.DataSource;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.support.JdbcTransactionManager;
import org.springframework.transaction.TransactionManager;

@Slf4j
//@Configuration
public class DbConfig {

  @Bean
  public DataSource dataSource() {
    log.info("datasource bean create");
    HikariDataSource dataSource = new HikariDataSource();
    dataSource.setDriverClassName("org.h2.Driver");
    dataSource.setJdbcUrl("jdbc:h2:mem:test");
    dataSource.setUsername("sa");
    dataSource.setPassword("");

    return dataSource;
  }

  @Bean
  public TransactionManager transactionManager() {
    log.info("transactionManager bean create");
    return new JdbcTransactionManager(dataSource());
  }

  @Bean
  public JdbcTemplate jdbcTemplate() {
    log.info("jdbcTemplate bean create");
    return new JdbcTemplate(dataSource());
  }
}

File: spring-jpa/jpa-routing-datasource/jpa-routing-datasource/src/main/kotlin/com/toy/jparoutingdatasource/config/DbConfig.kt
package com.toy.jparoutingdatasource.config

import com.zaxxer.hikari.HikariDataSource
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.boot.autoconfigure.EnableAutoConfiguration
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean
import org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.boot.jdbc.DataSourceBuilder
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary
import org.springframework.jdbc.datasource.LazyConnectionDataSourceProxy
import org.springframework.jdbc.datasource.lookup.AbstractRoutingDataSource
import org.springframework.transaction.annotation.EnableTransactionManagement
import org.springframework.transaction.support.TransactionSynchronizationManager
import javax.sql.DataSource

@Configuration
@EnableTransactionManagement
@EnableAutoConfiguration(exclude = [DataSourceAutoConfiguration::class])
class DbConfig {

  @Bean(name = ["masterDataSource"])
  @ConfigurationProperties(prefix = "spring.datasource.master.hikari")
  fun masterDataSource(): HikariDataSource = DataSourceBuilder.create()
    .type(HikariDataSource::class.java)
    .build()

  @Bean(name = ["slaveDataSource"])
  @ConfigurationProperties(prefix = "spring.datasource.slave.hikari")
  fun slaveDataSource(): HikariDataSource = DataSourceBuilder.create()
    .type(HikariDataSource::class.java)
    .build()
    .apply { isReadOnly = true }

  @Bean
  @ConditionalOnBean(name = ["masterDataSource", "slaveDataSource"])
  fun routingDataSource(
    @Qualifier("masterDataSource") masterDataSource: DataSource,
    @Qualifier("slaveDataSource") slaveDataSource: DataSource
  ): DataSource {
    val routingDataSource = RoutingDataSource()
    val dataSources: Map<Any, Any> = mapOf(
      "master" to masterDataSource,
      "slave" to slaveDataSource
    )
    routingDataSource.setTargetDataSources(dataSources)
    routingDataSource.setDefaultTargetDataSource(masterDataSource)
    return routingDataSource
  }

  @Primary
  @Bean(name = ["currentDataSource"])
  @ConditionalOnBean(name = ["routingDataSource"])
  fun currentDataSource(routingDataSource: DataSource) = LazyConnectionDataSourceProxy(routingDataSource)
}

class RoutingDataSource: AbstractRoutingDataSource() {
  override fun determineCurrentLookupKey() = when {
    TransactionSynchronizationManager.isCurrentTransactionReadOnly() -> "slave"
    else -> "master"
  }
}
